chore: update documentation (#5)
camptocamp/devops-stack-module-backup: README.adoc

A https://devops-stack.io[DevOps Stack] module to deploy a backup solution based on https://velero.io[Velero].

This module add the ability to perform backups of a cluster. This includes all Kubernetes objects, although with the GitOps approach of the Devops Stack, it is not necessary to have a backup of all Kubernetes objects, as they are generally disposable and recreated by ArgoCD.

The Velero chart used by this module is shipped in this repository as well, in order to avoid any unwanted behaviors caused by unsupported versions.

[cols="1,1,1",options="autowidth,header"]
|===
|Current Chart Version |Original Repository |Default Values
|*{velero-chart-version}* |{original-repo-url}/tree/main/charts/velero[Chart] | https://artifacthub.io/packages/helm/vmware-tanzu/velero/{velero-chart-version}?modal=values[`values.yaml`]
|===

== Usage

=== Backup

Here is an example of the module usage:

- A default backend in a AWS S3 bucket, accessed with a ServiceAccount
- A schedule at 4 a.m, which uses EBS snapshots as backend for PVC
- A schedule at 2 a.m, which uses Restic as backend for PVC
- Prometheus monitoring with a Grafana dashboard

[source,terraform]
----

> module "backup" {

Reviewer: I would rename this to `velero`, since the module is no longer called `backup`.

```suggestion
module "velero" {
```

>   source           = "git::https://github.com/camptocamp/devops-stack-module-velero.git//eks"

Reviewer: ```suggestion
  source           = "git::https://github.com/camptocamp/devops-stack-module-velero.git//eks?ref=<RELEASE>""
```

  cluster_name     = local.cluster_name
  base_domain      = local.base_domain
  argocd_namespace = module.argocd_bootstrap.argocd_namespace

>   ``

Reviewer: Is this line desired or just an artifact? :)

  enable_monitoring_dashboard = true

  # Automated backups can be specified.
  #
  # The included Resources is very important: if the resource `pods` is specified,
  # then velero will not attempt to use the native snapshot feature of the object
  # storage of the cloud provider (EBS snapshots), and will use the File System Backup, i.e

>   # Restic.

Reviewer: I would put this in an admonition after the example. Maybe it's me, but I think this warning would be more "visible" within an important box.

  backup_schedules = {

    # This schedule will use the native snapshot ability to backup the PVC
    snapshot-schedule = {

>       # Can temporarly disable snapshot schedules

Reviewer: ```suggestion
      # Can temporarily disable snapshot schedules
```

      disabled = false
      # Cron-style scheduling
      schedule = "2 4 * * *"
      template = {
        # Select the backend for this schedule
        storageLocation    = "backup-bucket"
        includedNamespaces = ["wordpress"]
        includedResources  = ["persistentVolumes", "persistentVolumeClaims"]
      }
    },
    # This schedule will force the Restic backend to backup the PVC
    restic-schedule = {
      disabled = false
      schedule = "2 2 * * *"
      template = {

>         # If using an non-default location:

Reviewer: It's not clear to me what this means. Maybe it's just me not being familiarized with the module behaviour.

= devops-stack-module-velero
// Document attributes to replace along the document
// Here you can define variables for something that keeps repeating along the text

> :chart-version: 5.0.2

Reviewer: ```suggestion
:velero-chart-version: 5.0.2
```

Reviewer: @tyrossel You need to change this attribute, so the automatic updater also changes the attribute on the README.adoc. The prefix of the attribute is the same name as the name of the dependency on the Chart.yaml.

:original-repo-url: https://github.com/vmware-tanzu/helm-charts

A https://devops-stack.io[DevOps Stack] module to deploy a backup solution based on https://velero.io[Velero].

This module add the ability to perform backups of a cluster. This includes all Kubernetes objects, although

The Velero chart used by this module is shipped in this repository as well, in order to avoid any unwanted behaviors caused by unsupported versions.

[cols="1,1,1",options="autowidth,header"]
|===
|Current Chart Version |Original Repository |Default Values

> |*{chart-version}* |{original-repo-url}/tree/main/charts/velero[Chart] | https://artifacthub.io/packages/helm/vmware-tanzu/velero/{chart-version}?modal=values[`values.yaml`]

Reviewer: ```suggestion
|*{velero-chart-version}* |{original-repo-url}/tree/main/charts/velero[Chart] | https://artifacthub.io/packages/helm/vmware-tanzu/velero/{velero-chart-version}?modal=values[`values.yaml`]
```

|===

== Usage

=== Backup

Here is an example of the module usage in the
_Here you will show examples of how to call this module on your code. Note that the `depends_on` needs to be adapted to your use case._

_You can start by giving a simple example and then show some others with increasing complexity and/or for specific use cases._

[source,terraform]
----
module "backup" {

>   source           = "git::https://github.com/camptocamp/devops-stack-module-backup.git//eks"

Reviewer: ```suggestion
  source           = "git::https://github.com/camptocamp/devops-stack-module-velero.git//eks"
```

    iam_role_arn = module.iam_assumable_role_velero.iam_role_arn
  }

  dependency_ids = {
    kube-prometheus-stack = module.kube-prometheus-stack.id,
    argocd                = module.argocd_bootstrap.id # TODO Can cause issues on the first cluster bootstrap. If it's the case, deploy after the last argocd.

>   }

Reviewer: Did you intend to remove the `dependency_ids`?

Author: I originally thought that it wasn't really part of the module usage, but more like a deployment config. But now I agree that it should also be documented in this example.

>   # Triggers an alert if too many backups are partially failed (ratio)
>   alert_partially_failed_ratio = 0.25
>   # Triggers an alert if too many backups are failed (ratio)
>   alert_failed_ratio = 0.1

Reviewer: ```suggestion
  # Triggers an alert if too many backups are partially failed (ratio)
  alert_partially_failed_ratio = 0.25 # 25%
  # Triggers an alert if too many backups are failed (ratio)
  alert_failed_ratio = 0.1 # 10%
```

  # Triggers an alert if latest backup is older than this, in seconds
  alert_backup_timeout = 60 * 60 * 24

}

----

> Once the Velero controller is deployed on the server, it is also possible to perform manual backup from the `velero` client:

Reviewer: ```suggestion
Once the Velero controller is deployed on the server, it is also possible to perform a manual backup using the `velero` client:
```

Reviewer: Another question: have you documented somewhere else how this client authenticates to the `velero` controller inside the cluster?

[source,bash]
----
velero backup create \
    --included-namespaces wordpress,loki \
    --included-resources pv,pvc,pods \
    --storage-location my-s3-bucket \
    my-new-backup
----

=== Restore

The backup can be restored manually with the following commands:

[source,bash]
----
# Restore from a backup
velero restore create --from-backup my-new-backup
# Restore from the last backup of a schedule
velero restore create --from-schedule restic-schedule
----

More options are described in the https://velero.io/docs/main/restore-reference/[Velero documentation].

> NOTE: In order to perform restore successfully, Velero need that the object doesn't already exists in the target cluster.
> As a disaster recovery tool, it is not able to merge an existing object such as a PVC with its backup.

Reviewer: ```suggestion
[IMPORTANT]
====
In order to successfully perform a restore, Velero requires that the object doesn't already exists in the target cluster.
As a disaster recovery tool, it is not able to merge an existing object, such as a PVC, with its backup.
====
```

== Monitoring

> If enabled, this module exposes Velero metrics and setup a scrape target for Prometheus.

Reviewer: ```suggestion
If enabled, this module exposes Velero metrics and setups a scrape target for Prometheus.
```

=== Grafana dashboard

> This module contains a Grafana dashboard to monitor the frequency, state and frequency of backups, as shown here below.

Reviewer: ```suggestion
This module contains a Grafana dashboard to monitor the state and frequency of backups, as shown here below.
```

> image:https://raw.githubusercontent.com/camptocamp/devops-stack-module-velero/doc/docs/modules/ROOT/assets/images/grafana-velero-dashboard.png["Grafana dashboard with backup informations",link=https://raw.githubusercontent.com/camptocamp/devops-stack-module-velero/doc/docs/modules/ROOT/assets/images/grafana-velero-dashboard.png]

Reviewer: ```suggestion
image:https://raw.githubusercontent.com/camptocamp/devops-stack-module-velero/doc/docs/modules/ROOT/assets/images/grafana-velero-dashboard.png["Grafana dashboard with backup informations",link=https://raw.githubusercontent.com/camptocamp/devops-stack-module-velero/doc/docs/modules/ROOT/assets/images/grafana-velero-dashboard.png, window=_blank]
```

=== Alerts

> This module also setup some alerting in Prometheus for the following cases:

Reviewer: ```suggestion
This module also setups some alerts in Alertmanager for the following cases:
```

> - Partially failed backups ratio higher than a specified amount (`alert_partially_failed_ratio`)
> - Failed backups ratio higher than a specified amount (`alert_failed_ratio`)
> - No successful backup for a specified amount of time (`alert_backup_timeout`)

Reviewer: ```suggestion
- Partially failed backups ratio higher than a specified amount (`alert_partially_failed_ratio`);
- Failed backups ratio higher than a specified amount (`alert_failed_ratio`);
- No successful backup for a specified amount of time (`alert_backup_timeout`);
```

== Limitations

=== KinD setup

> Velero cannot perform backups of `hostPath` volumes, which are the volumes used in a KinD setup. Therefore, this module cannot be used in a Devops stack on KinD.

Reviewer: ```suggestion
Velero cannot perform backups of `hostPath` volumes, which are the volumes used in a KinD platform. Therefore, this module cannot be used in a DevOps Stack on KinD.
```

== Technical Reference

=== Dependencies

==== `module.kube-prometheus-stack`

As this module contains a Prometheus metrics endpoint and a Grafana dashboard, it has to be deployed after the https://github.com/camptocamp/devops-stack-module-kube-prometheus-stack[Prometheus stack].

==== `module.argocd`

Velero is deployed as an ArgoCD application, which means that ArgoCD has to be deployed first.

==== S3 bucket or equivalent

In order to perform backups, Velero needs a storage location for backups, either a S3 compatible storage (AWS S3, Minio, etc.), Azure Blob storage, etc.

> Velero backs all Kubernetes objects to the designated

Reviewer: I would put this block somewhere else, we can discuss together if you want.

The `Dependencies` section was designed more thinking about which other Terraform modules this module depends on.
